src/meta/src/manager/catalog/mod.rs

```diff
@@ -481,67 +481,117 @@ where
         &self,
         table_id: TableId,
         internal_table_ids: Vec<TableId>,
+        index_and_table_ids: Vec<(IndexId, TableId)>,
     ) -> MetaResult<NotificationVersion> {
         let core = &mut *self.core.lock().await;
         let database_core = &mut core.database;
         let user_core = &mut core.user;
         let table = Table::select(self.env.meta_store(), &table_id).await?;
         if let Some(table) = table {
-            match database_core.get_ref_count(table_id) {
-                Some(ref_count) => Err(MetaError::permission_denied(format!(
-                    "Fail to delete table `{}` because {} other relation(s) depend on it",
-                    table.name, ref_count
-                ))),
-                None => {
-                    let dependent_relations = table.dependent_relations.clone();
-                    let mut transaction = Transaction::default();
+            if let Some(ref_count) = database_core.get_ref_count(table_id) {
+                if ref_count > index_and_table_ids.len() {
+                    return Err(MetaError::permission_denied(format!(
+                        "Fail to delete table `{}` because {} other relation(s) depend on it",
+                        table.name, ref_count
+                    )));
+                }
+            }
 
-                    let mut tables_to_drop =
-                        future::join_all(internal_table_ids.into_iter().map(|id| async move {
-                            Table::select(self.env.meta_store(), &id).await
-                        }))
-                        .await
-                        .into_iter()
-                        .map_ok(|table| table.unwrap())
-                        .collect::<MetadataModelResult<Vec<_>>>()?;
-                    tables_to_drop.push(table);
-
-                    for table in &tables_to_drop {
-                        table.delete_in_transaction(&mut transaction)?;
+            let dependent_relations = table.dependent_relations.clone();
+            let mut transaction = Transaction::default();
+
+            let mut indexes_post_work_vec = vec![];
+            // Delete indexes
+            for (index_id, index_table_id) in index_and_table_ids {
+                let index = Index::select(self.env.meta_store(), &index_id).await?;
+                if let Some(index) = index {
+                    index.delete_in_transaction(&mut transaction)?;
+                    assert_eq!(index_table_id, index.index_table_id);
+
+                    // drop index table
+                    let table = Table::select(self.env.meta_store(), &index_table_id).await?;
+                    if let Some(table) = table {
+                        match database_core.get_ref_count(index_table_id) {
```

Comment:
Non-related to this PR, do we support or going to support creating index/mv on index now? 🤔

Reply:
Create an index on an index is forbidden, however, I think creating a mv on an index is ok (considering the delta join case).

Reply:
Okay, then this check is still needed. Good job 👍🏻, I think this PR is LGTM now.